### PR TITLE
UIOR-234 Require order approval to "open" order

### PR DESCRIPTION
**Pull request by Feboll**
## Purpose

https://issues.folio.org/browse/UIOR-234
Some institutions require that orders are approved before they can be placed.

## Approach

Add button for approve order and add approval date to view

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/components/PurchaseOrder/PO.js [hunk 1]
```diff
@@ -204,6 +205,20 @@ class PO extends Component {
...
+    const { showToast, mutator, resources } = this.props;
+    const order = get(resources, ['order', 'records', 0]);
+    const orderNumber = get(order, 'poNumber', '');
+
+    try {
+      await updateOrderResource(order, mutator.order, { approved: true });
+    } catch (error) {
+      await showUpdateOrderError(e, this.callout, this.openOrderErrorModalShow);
+    } finally {
+      showToast('ui-orders.order.approved.success', 'success', { orderNumber });
```
Comment (NikitaSedyx) [inline src/components/PurchaseOrder/PO.js]: success in finally ?

File: src/components/PurchaseOrder/PO.js [hunk 2]
```diff
@@ -316,12 +331,16 @@ class PO extends Component {
       stripes,
     } = this.props;
     const order = this.getOrder();
+    const isApproved = get(order, 'approved');
+    const approvalsSetting = getOrderApprovalsSetting(get(parentResources, 'approvalsSetting.records', {}));
+    const { isApprovalRequired } = approvalsSetting;
```
Comment (NikitaSedyx) [inline src/components/PurchaseOrder/PO.js]: `const { isApprovalRequired } = getOrderApprovalsSetting(get(parentResources, 'approvalsSetting.records', {}));`?

File: src/components/PurchaseOrder/PO.js [hunk 3]
```diff
@@ -316,12 +331,16 @@ class PO extends Component {
...
     const closingReasons = get(parentResources, 'closingReasons.records', []);
     const orderNumber = get(order, 'poNumber', '');
     const poLines = get(order, 'compositePoLines', []);
     const workflowStatus = get(order, 'workflowStatus');
     const isCloseOrderButtonVisible = workflowStatus === WORKFLOW_STATUS.open;
-    const isOpenOrderButtonVisible = workflowStatus === WORKFLOW_STATUS.pending && poLines.length > 0;
+    const workflowIsPending = workflowStatus === WORKFLOW_STATUS.pending;
```
Comment (NikitaSedyx) [inline src/components/PurchaseOrder/PO.js]: is it used somewhere else?

File: src/components/PurchaseOrder/PO.js [hunk 4]
```diff
@@ -316,12 +331,16 @@ class PO extends Component {
...
+    const isOpenOrderButtonVisible = workflowIsPending && poLines.length > 0 && (isApproved || !isApprovalRequired);
```
Comment (NikitaSedyx) [inline src/components/PurchaseOrder/PO.js]: seems we need a function for it

File: src/components/PurchaseOrder/PO.js [hunk 5]
```diff
@@ -204,6 +206,19 @@ class PO extends Component {
     this.unmountCloseOrderModal();
   };
 
+  approveOrder = async (e) => {
+    const { showToast, mutator, resources } = this.props;
+    const order = get(resources, ['order', 'records', 0]);
```
Comment (aliaksei-chumakou) [inline src/components/PurchaseOrder/PO.js]: `const order = this.getOrder();` ?

File: src/components/PurchaseOrder/PO.js [hunk 6]
```diff
@@ -204,6 +206,19 @@ class PO extends Component {
...
+    const orderNumber = get(order, 'poNumber', '');
+
+    try {
+      await updateOrderResource(order, mutator.order, { approved: true });
+      showToast('ui-orders.order.approved.success', 'success', { orderNumber });
+    } catch (error) {
+      await showUpdateOrderError(e, this.callout, this.openOrderErrorModalShow);
```
Comment (aliaksei-chumakou) [inline src/components/PurchaseOrder/PO.js]: Have you checked this working? I don't see `e` is defined here

File: src/components/PurchaseOrder/PO.js [hunk 7]
```diff
@@ -405,6 +423,20 @@ class PO extends Component {
           </IfPermission>
 
           <IfPermission perm="orders.item.put">
+            <IfPermission perm="orders.item.approve">
+              {isApprovalRequired && !isApproved && (
```
Comment (aliaksei-chumakou) [inline src/components/PurchaseOrder/PO.js]: `isApprovalRequired && !isApproved` looks confusing, could you please leverage this to one boolean?

File: src/components/PurchaseOrder/PO.js (hunk 6 above)
Comment (aliaksei-chumakou) [reply src/components/PurchaseOrder/PO.js]: ok, it's defined but looks very confusing. Usually it's an error in catch block. Could you please give it a proper name?

File: src/components/PurchaseOrder/PO.js [hunk 8]
```diff
@@ -50,11 +50,13 @@ import { RenewalsView } from './renewals';
 import LinesLimit from './LinesLimit';
 import POInvoicesContainer from './POInvoices';
 import {
+  isOpenAvailableForOrder,
   isReceiveAvailableForOrder,
 } from './util';
 import { UpdateOrderErrorModal } from './UpdateOrderErrorModal';
 
 import css from './PO.css';
+import { getOrderApprovalsSetting } from '../../common/utils/getOrderApprovalsSetting';
```
Comment (aliaksei-chumakou) [inline src/components/PurchaseOrder/PO.js]: shouldn't it be above, around `getAddresses` imported `from '../../common/utils';`

File: translations/ui-orders/en_US.json [hunk 9]
```diff
@@ -727,4 +727,4 @@
     "relatedInvoices.vendorName": "Vendor name",
     "settings.approvals": "Approvals",
     "settings.approvals.isApprovalRequired": "\"Approval required\" to open orders"
-}
\ No newline at end of file
+}
```
Comment (AndreiShumski2) [inline translations/ui-orders/en_US.json]: no needed blank line

**Pull request merged**